README.md
[![Moov Banner Logo](https://user-images.githubusercontent.com/20115216/104214617-885b3c80-53ec-11eb-8ce0-9fc745fb5bfc.png)](https://github.com/moov-io)

<!--
<p align="center">
  <a href="https://moov-io.github.io/ach/">Project Documentation</a>
  ·
  <a href="https://moov-io.github.io/ach/api/#get-/files">API Endpoints</a>
  ·
  <a href="https://moov.io/blog/education/ach-api-guide/">API Guide</a>
  ·
  <a href="https://slack.moov.io/">Community</a>
  ·
  <a href="https://moov.io/blog/">Blog</a>
  <br>
  <br>
</p>
-->

[![GoDoc](https://godoc.org/github.com/moov-io/fips-state-codes?status.svg)](https://godoc.org/github.com/moov-io/fips-state-codes)
[![Build Status](https://github.com/moov-io/fips-state-codes/workflows/Go/badge.svg)](https://github.com/moov-io/fips-state-codes/actions)
[![Coverage Status](https://codecov.io/gh/moov-io/fips-state-codes/branch/master/graph/badge.svg)](https://codecov.io/gh/moov-io/fips-state-codes)
[![Go Report Card](https://goreportcard.com/badge/github.com/moov-io/fips-state-codes)](https://goreportcard.com/report/github.com/moov-io/fips-state-codes)
[![Repo Size](https://img.shields.io/github/languages/code-size/moov-io/fips-state-codes?label=project%20size)](https://github.com/moov-io/fips-state-codes)
[![Apache 2 License](https://img.shields.io/badge/license-Apache2-blue.svg)](https://raw.githubusercontent.com/moov-io/fips-state-codes/master/LICENSE)
[![Slack Channel](https://slack.moov.io/badge.svg?bg=e01563&fgColor=fffff)](https://slack.moov.io/)
[![GitHub Stars](https://img.shields.io/github/stars/moov-io/fips-state-codes)](https://github.com/moov-io/fips-state-codes)
[![Twitter](https://img.shields.io/twitter/follow/moov?style=social)](https://twitter.com/moov?lang=en)

# moov-io/fips-state-codes
Moov's mission is to give developers an easy way to create and integrate bank processing into their own software products. Our open source projects are each focused on solving a single responsibility in financial services and designed around performance, scalability, and ease of use.

fips-state-codes provides a mapping function to translate standard US based 2-alphanumeric state codes into FIPS 2-digit state codes.

## Project Status

fips-state-codes is included in multiple open-source projects Moov offers and is used in production environments. Please star the project if you are interested in its progress. If you find any bugs or desire additional encryption/encoding algorithms we would appreciate an issue or pull request. Thanks!

## Usage

Call the FIPSCodeFromStateCode function to translate a state code to a FIPS state code.

**Translate**
```go
// fipsStateCode contains value "17" in below example
fipsStateCode, err := fips_state_codes.FIPSCodeFromStateCode("IL")
if err != nil {
    // do something
}
```

## Getting help

channel | info
 ------- | -------
Twitter [@moov](https://twitter.com/moov)	| You can follow Moov.io's Twitter feed to get updates on our project(s). You can also tweet us questions or just share blogs or stories.
[GitHub Issue](https://github.com/moov-io/fips-state-codes/issues/new) | If you are able to reproduce a problem please open a GitHub Issue under the specific project that caused the error.
[moov-io slack](https://slack.moov.io/) | Join our slack channel to have an interactive discussion about the development of the project.

## Supported and tested platforms

- 64-bit Linux (Ubuntu, Debian), macOS, and Windows

## Contributing

Yes please! Please review our [Contributing guide](CONTRIBUTING.md) and [Code of Conduct](https://github.com/moov-io/ach/blob/master/CODE_OF_CONDUCT.md) to get started! Checkout our [issues for first time contributors](https://github.com/moov-io/watchman/contribute) for something to help out with.

This project uses [Go Modules](https://go.dev/blog/using-go-modules) and Go v1.18 or newer. See [Golang's install instructions](https://golang.org/doc/install) for help setting up Go. You can download the source code and we offer [tagged and released versions](https://github.com/moov-io/ach/releases/latest) as well. We highly recommend you use a tagged release for production.

## License

Apache License 2.0 - See [LICENSE](LICENSE) for details.


CHANGELOG.md
## v0.1.0 (Release 2023-07-24)

Add DC and other outlying area codes

## v0.0.1 (Released 2022-08-23)

Initial release


code_mapper_test.go
// Licensed to The Moov Authors under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. The Moov Authors licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fips_state_codes_test

import (
	"fmt"
	"testing"

	fips_state_codes "github.com/moov-io/fips-state-codes"
	"github.com/stretchr/testify/require"
)

func TestFIPSCodeFromStateCode(t *testing.T) {
	result, err := fips_state_codes.FIPSCodeFromStateCode("IL")
	require.NoError(t, err)
	require.Equal(t, "17", result)
}

func TestFIPSCodeFromStateCode_ErrorsOnUnknownStateCode(t *testing.T) {
	inputCode := "ZZ"
	result, err := fips_state_codes.FIPSCodeFromStateCode(inputCode)
	require.Equal(t, "", result)
	require.EqualError(t, err, fmt.Sprintf("unknown state code: %s", inputCode))
}

func TestStateCodeFromFIPSCode(t *testing.T) {
	result, err := fips_state_codes.StateCodeFromFIPSCode("17")
	require.NoError(t, err)
	require.Equal(t, "IL", result)
}

func TestStateCodeFromFIPSCode_ErrorsOnUnknownStateCode(t *testing.T) {
	inputCode := "ZZ"
	result, err := fips_state_codes.StateCodeFromFIPSCode(inputCode)
	require.Equal(t, "", result)
	require.EqualError(t, err, fmt.Sprintf("unknown fips code: %s", inputCode))
}


state_to_fips_code_mapper_test.go
// Licensed to The Moov Authors under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. The Moov Authors licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fips_state_codes_test

import (
	"fmt"
	fips_state_codes "github.com/moov-io/fips-state-codes"
	"github.com/stretchr/testify/require"
	"testing"
)

func TestFIPSCodeFromStateCode(t *testing.T) {
	result, err := fips_state_codes.FIPSCodeFromStateCode("IL")
	require.NoError(t, err)
	require.Equal(t, "17", result)
}

func TestFIPSCodeFromStateCode_ErrorsOnUnknownStateCode(t *testing.T) {
	inputCode := "ZZ"
	result, err := fips_state_codes.FIPSCodeFromStateCode(inputCode)
	require.Equal(t, "", result)
	require.Error(t, err)
	require.Equal(t, fmt.Sprintf("unknown state code: %s", inputCode), err.Error())
}


state_to_fips_code_mapper.go
// Licensed to The Moov Authors under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. The Moov Authors licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fips_state_codes

import (
	"fmt"
)

var stateCodeToVIPSCodeMap = map[string]string{
	"AL": "01",
	"AK": "02",
	"AZ": "04",
	"AR": "05",
	"CA": "06",
	"CO": "08",
	"CT": "09",
	"DE": "10",
	"FL": "12",
	"GA": "13",
	"HI": "15",
	"ID": "16",
	"IL": "17",
	"IN": "18",
	"IA": "19",
	"KS": "20",
	"KY": "21",
	"LA": "22",
	"ME": "23",
	"MD": "24",
	"MA": "25",
	"MI": "26",
	"MN": "27",
	"MS": "28",
	"MO": "29",
	"MT": "30",
	"NE": "31",
	"NV": "32",
	"NH": "33",
	"NJ": "34",
	"NM": "35",
	"NY": "36",
	"NC": "37",
	"ND": "38",
	"OH": "39",
	"OK": "40",
	"OR": "41",
	"PA": "42",
	"RI": "44",
	"SC": "45",
	"SD": "46",
	"TN": "47",
	"TX": "48",
	"UT": "49",
	"VT": "50",
	"VA": "51",
	"WA": "53",
	"WV": "54",
	"WI": "55",
	"WY": "56",
	"AS": "60",
	"GU": "66",
	"MP": "69",
	"PR": "72",
	"VI": "78",
}

func FIPSCodeFromStateCode(stateCode string) (string, error) {
	got, ok := stateCodeToVIPSCodeMap[stateCode]
	if !ok {
		return "", fmt.Errorf("unknown state code: %s", stateCode)
	}
	return got, nil
}


code_mapper.go
// Licensed to The Moov Authors under one or more contributor
// license agreements. See the NOTICE file distributed with
// this work for additional information regarding copyright
// ownership. The Moov Authors licenses this file to you under
// the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fips_state_codes

import (
	"fmt"
)

var stateCodeToFIPSCodeMap = map[string]string{
	"AL": "01",
	"AK": "02",
	"AZ": "04",
	"AR": "05",
	"CA": "06",
	"CO": "08",
	"CT": "09",
	"DE": "10",
	"DC": "11",
	"FL": "12",
	"GA": "13",
	"HI": "15",
	"ID": "16",
	"IL": "17",
	"IN": "18",
	"IA": "19",
	"KS": "20",
	"KY": "21",
	"LA": "22",
	"ME": "23",
	"MD": "24",
	"MA": "25",
	"MI": "26",
	"MN": "27",
	"MS": "28",
	"MO": "29",
	"MT": "30",
	"NE": "31",
	"NV": "32",
	"NH": "33",
	"NJ": "34",
	"NM": "35",
	"NY": "36",
	"NC": "37",
	"ND": "38",
	"OH": "39",
	"OK": "40",
	"OR": "41",
	"PA": "42",
	"RI": "44",
	"SC": "45",
	"SD": "46",
	"TN": "47",
	"TX": "48",
	"UT": "49",
	"VT": "50",
	"VA": "51",
	"WA": "53",
	"WV": "54",
	"WI": "55",
	"WY": "56",
	"AS": "60",
	"FM": "64",
	"GU": "66",
	"MH": "68",
	"MP": "69",
	"PW": "70",
	"PR": "72",
	"UM": "74",
	"VI": "78",
}

func FIPSCodeFromStateCode(stateCode string) (string, error) {
	got, ok := stateCodeToFIPSCodeMap[stateCode]
	if !ok {
		return "", fmt.Errorf("unknown state code: %s", stateCode)
	}
	return got, nil
}

var fipsCodeToStateCodeMap = map[string]string{
	"01": "AL",
	"02": "AK",
	"04": "AZ",
	"05": "AR",
	"06": "CA",
	"08": "CO",
	"09": "CT",
	"10": "DE",
	"11": "DC",
	"12": "FL",
	"13": "GA",
	"15": "HI",
	"16": "ID",
	"17": "IL",
	"18": "IN",
	"19": "IA",
	"20": "KS",
	"21": "KY",
	"22": "LA",
	"23": "ME",
	"24": "MD",
	"25": "MA",
	"26": "MI",
	"27": "MN",
	"28": "MS",
	"29": "MO",
	"30": "MT",
	"31": "NE",
	"32": "NV",
	"33": "NH",
	"34": "NJ",
	"35": "NM",
	"36": "NY",
	"37": "NC",
	"38": "ND",
	"39": "OH",
	"40": "OK",
	"41": "OR",
	"42": "PA",
	"44": "RI",
	"45": "SC",
	"46": "SD",
	"47": "TN",
	"48": "TX",
	"49": "UT",
	"50": "VT",
	"51": "VA",
	"53": "WA",
	"54": "WV",
	"55": "WI",
	"56": "WY",
	"60": "AS",
	"64": "FM",
	"66": "GU",
	"68": "MH",
	"69": "MP",
	"70": "PW",
	"72": "PR",
	"74": "UM",
	"78": "VI",
}

func StateCodeFromFIPSCode(fipsCode string) (string, error) {
	got, ok := fipsCodeToStateCodeMap[fipsCode]
	if !ok {
		return "", fmt.Errorf("unknown fips code: %s", fipsCode)
	}
	return got, nil
}
